x/market/types/message_deal.go

```diff
@@ -71,3 +71,85 @@ func (msg *MsgCreateDeal) GetSigners() []sdk.AccAddress {
 	}
 	return []sdk.AccAddress{creator}
 }
+
+var _ sdk.Msg = &MsgSellData{}
+
+func NewMsgSellData(cert DataValidationCertificate, seller string) *MsgSellData {
+	return &MsgSellData{
+		Cert:   &cert,
+		Seller: seller,
+	}
+}
+
+func (msg *MsgSellData) Route() string {
+	return RouterKey
+}
+
+func (msg *MsgSellData) Type() string {
+	return "SellData"
+}
+
+// ValidateBasic is validation for MsgSellData.
+func (msg *MsgSellData) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.Seller)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
+	}
+
+	if msg.Cert == nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "empty certificate")
+	}
+
+	signature := msg.Cert.Signature
+	if signature == nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "empty signature")
+	}
+
+	unsignedCert := msg.Cert.UnsignedCert
+	if unsignedCert == nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "no unsigned data certificate")
+	}
+
+	if unsignedCert.DealId <= 0 {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid deal id format")
+	}
+
+	if unsignedCert.DataHash == "" {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "empty data hash")
+	}
+
+	if unsignedCert.EncryptedDataUrl == "" {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "empty encrypted data url")
+	}
+
+	if unsignedCert.DataValidatorAddress == "" {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "empty data validator address")
+	}
+	_, err = sdk.AccAddressFromBech32(unsignedCert.DataValidatorAddress)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid data validator address (%s)", err)
+	}
+
+	if unsignedCert.RequesterAddress == "" {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "empty requester address")
+	}
+	_, err = sdk.AccAddressFromBech32(unsignedCert.RequesterAddress)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid requester address (%s)", err)
+	}
+
+	return nil
+}
```

Comment:
We need to check if the seller and requester address (in unsigned certificate) are the same.
It may be missed and I think this check would be done here `ValidateBasic`. wdyt?

Reply:
Yes, you're right. I need to add it.